### PR TITLE
Add battlepass

### PR DESCRIPTION
**Pull request by Fizo55**

### CODE REVIEW

File: src/NosCore.Core/Configuration/WorldConfiguration.cs [hunk 1]
```diff
@@ -72,6 +73,18 @@ public class WorldConfiguration : ServerConfiguration
 #pragma warning disable CA2227 // Collection properties should be read only
         public Dictionary<string, List<BasicEquipment>> BasicEquipments { get; set; } = new();
 #pragma warning restore CA2227 // Collection properties should be read only
+
+        [Required]
+        public Battlepass Battlepass { get; set; } = null!;
```
Comment (erwan-joly) [inline src/NosCore.Core/Configuration/WorldConfiguration.cs:78]: BattlepassConfiguration

File: src/NosCore.Core/Configuration/WorldConfiguration.cs [hunk 2]
```diff
@@ -72,6 +73,18 @@ public class WorldConfiguration : ServerConfiguration
...
+    public class Battlepass
+    {
+        public bool IsBattlePassIconEnabled { get; set; }
+
+        public int MaxBattlePassPoints { get; set; }
+
+        public Instant EndSeason { get; set; }
     }
 
     public class BasicEquipment
```
Comment (erwan-joly) [inline src/NosCore.Core/Configuration/WorldConfiguration.cs:90]: seems like i've been lazy here

File: src/NosCore.Core/Configuration/WorldConfiguration.cs [hunk 3]
```diff
@@ -72,6 +73,18 @@ public class WorldConfiguration : ServerConfiguration
 #pragma warning disable CA2227 // Collection properties should be read only
         public Dictionary<string, List<BasicEquipment>> BasicEquipments { get; set; } = new();
 #pragma warning restore CA2227 // Collection properties should be read only
+
+        [Required]
+        public Battlepass Battlepass { get; set; } = null!;
+    }
+
+    public class Battlepass
```
Comment (erwan-joly) [inline src/NosCore.Core/Configuration/WorldConfiguration.cs:81]: need to move in a file

File: src/NosCore.Database/Entities/Quest.cs [hunk 4]
```diff
@@ -43,6 +43,8 @@ public Quest()
 
         public QuestType QuestType { get; set; }
 
+        public MissionType? QuestSubType { get; set; }
```
Comment (erwan-joly) [inline src/NosCore.Database/Entities/Quest.cs:46]: what is that ?

File: src/NosCore.Database/Migrations/NosCoreContextModelSnapshot.cs [hunk 5]
```diff
@@ -32,6 +32,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
             NpgsqlModelBuilderExtensions.HasPostgresEnum(modelBuilder, "family_authority_type", new[] { "none", "put", "all" });
             NpgsqlModelBuilderExtensions.HasPostgresEnum(modelBuilder, "family_log_type", new[] { "daily_message", "raid_won", "rainbow_battle", "family_xp", "family_level_up", "level_up", "item_upgraded", "right_changed", "authority_changed", "family_managed", "user_managed", "ware_house_added", "ware_house_removed" });
             NpgsqlModelBuilderExtensions.HasPostgresEnum(modelBuilder, "family_member_rank", new[] { "nothing", "old_uncle", "old_aunt", "father", "mother", "uncle", "aunt", "brother", "sister", "spouse", "brother2", "sister2", "old_son", "old_daugter", "middle_son", "middle_daughter", "young_son", "young_daugter", "old_little_son", "old_little_daughter", "little_son", "little_daughter", "middle_little_son", "middle_little_daugter" });
+            NpgsqlModelBuilderExtensions.HasPostgresEnum(modelBuilder, "frequency_type", new[] { "daily", "weekly", "seasonal" });
```
Comment (erwan-joly) [inline src/NosCore.Database/Migrations/NosCoreContextModelSnapshot.cs:35]: one_off sounds better

File: src/NosCore.Core/Configuration/WorldConfiguration.cs [hunk 6]
```diff
@@ -72,6 +73,18 @@ public class WorldConfiguration : ServerConfiguration
 #pragma warning disable CA2227 // Collection properties should be read only
         public Dictionary<string, List<BasicEquipment>> BasicEquipments { get; set; } = new();
 #pragma warning restore CA2227 // Collection properties should be read only
+
+        [Required]
+        public BattlepassConfiguration BattlepassConfiguration { get; set; } = null!;
+    }
+
+    public class BattlepassConfiguration
```
Comment (erwan-joly) [inline src/NosCore.Core/Configuration/WorldConfiguration.cs:81]: please move to it's own file

File: src/NosCore.GameObject/ComponentEntities/Extensions/CharacterEntityExtension.cs [hunk 7]
```diff
@@ -700,5 +701,39 @@ public static CInfoPacket GenerateCInfo(this ICharacterEntity visualEntity)
                 ArenaWinner = false
             };
         }
+
+        public static BptPacket GenerateBpt(this ICharacterEntity _, IClock clock, IOptions<WorldConfiguration> worldConfig)
```
Comment (erwan-joly) [inline src/NosCore.GameObject/ComponentEntities/Extensions/CharacterEntityExtension.cs:705]: this is not linked in anyway to a character so it shouldn't be in this file. Pretty sure it's sent only once so should be directly there

File: src/NosCore.GameObject/ComponentEntities/Extensions/CharacterEntityExtension.cs [hunk 8]
```diff
@@ -700,5 +701,39 @@ public static CInfoPacket GenerateCInfo(this ICharacterEntity visualEntity)
...
+            return new BptPacket
+            {
+                MinutesUntilSeasonEnd = (long)duration.TotalMinutes
+            };
+        }
+
+        public static BpmPacket GenerateBpm(this ICharacterEntity visualEntity, IOptions<WorldConfiguration> worldConfig)
+        {
+            List<BpmSubTypePacket> subPackets = new();
+            foreach (var quest in visualEntity.Quests.Values.Where(s => s.Quest.FrequencyType == FrequencyType.Daily || s.Quest.FrequencyType == FrequencyType.Daily)) // TODO : Improve that because the where condition isn't really true
```
Comment (erwan-joly) [inline src/NosCore.GameObject/ComponentEntities/Extensions/CharacterEntityExtension.cs:717]: daily and daily ?

File: src/NosCore.GameObject/Services/QuestService/QuestService.cs [hunk 9]
```diff
@@ -343,5 +343,31 @@ public async Task<bool> ValidateQuestAsync(ICharacterEntity character, short que
...
+            if (characterQuest.Quest.FrequencyType == FrequencyType.Daily)
+            {
+                daysToAdd = Duration.FromDays(1);
+            }
+            else if (characterQuest.Quest.FrequencyType == FrequencyType.Weekly)
+            {
+                daysToAdd = Duration.FromDays(7);
+            }
+
+            return Task.FromResult(Instant.Subtract(characterQuest.StartedOn.Plus(daysToAdd), _clock.GetCurrentInstant()).TotalMinutes);
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Services/QuestService/QuestService.cs:370]: so when you start a quest at 23:59 you have 1 day not 1 min ?

File: src/NosCore.GameObject/Services/QuestService/Quest.cs [hunk 10]
```diff
@@ -28,6 +32,27 @@ public class CharacterQuest : CharacterQuestDto
...
+            Instant instant = _worldConfiguration.Value.BattlepassConfiguration.EndSeason;
+
+            if (Quest.FrequencyType == FrequencyType.Daily)
+            {
+                instant = StartedOn.Plus(Duration.FromDays(1));
+            }
+            else if (Quest.FrequencyType == FrequencyType.Weekly)
+            {
+                instant = StartedOn.Plus(Duration.FromDays(7));
+            }
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Services/QuestService/Quest.cs:51]: use the new switch statement please

File: src/NosCore.PacketHandlers/Game/BpOpenPacketHandler.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            _quests = quests;
+        }
+
+        public override async Task ExecuteAsync(BpOpenPacket packet, ClientSession session)
+        {
+            await session.SendPacketAsync(session.Character.GenerateBpm(_clock, _worldConfiguration, _quests));
+            await session.SendPacketAsync(session.Character.GenerateBpp());
+            await session.SendPacketAsync(new BptPacket
+            {
+                MinutesUntilSeasonEnd = (long)Instant.Subtract(_worldConfiguration.Value.BattlepassConfiguration.EndSeason, _clock.GetCurrentInstant()).TotalMinutes
```
Comment (erwan-joly) [inline src/NosCore.PacketHandlers/Game/BpOpenPacketHandler.cs:33]: didn't noticed it was a server packet >< this should be part of the constructor

File: src/NosCore.GameObject/Services/QuestService/Quest.cs [hunk 12]
```diff
@@ -28,6 +32,23 @@ public class CharacterQuest : CharacterQuestDto
     {
         public Quest Quest { get; set; } = null!;
 
+        public double GetTotalMinutesLeftBeforeQuestEnd(IOptions<WorldConfiguration> worldConfiguration, IClock clock)
+        {
+            if (Quest == null)
+            {
+                return (double)0;
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Services/QuestService/Quest.cs:39]: 0D

File: src/NosCore.GameObject/Services/QuestService/QuestObjectives.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+namespace NosCore.GameObject.Services.QuestService
+{
+    public class QuestObjectives : QuestObjectiveDto
+    {
+        public int? GetMaxCount(QuestType questType)
+        {
+            return questType switch
+            {
+                QuestType.Hunt or QuestType.SpecialCollect or QuestType.TimesSpace or QuestType.Product => SecondData != null && ThirdData != null ? (int?)Math.Floor((double)((SecondData + ThirdData) / 2)) : SecondData == null ? ThirdData : SecondData,
+                QuestType.Brings or QuestType.CaptureWithoutGettingTheMonster or QuestType.Capture  => ThirdData != null && FourthData != null ? (int?)Math.Floor((double)((ThirdData + FourthData) / 2)) : ThirdData == null ? FourthData : ThirdData,
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Services/QuestService/QuestObjectives.cs:14]: This looks super weird should be added to the parser instead. I still don’t get what maxCount is

Comment (0Lucifer0) [reply src/NosCore.GameObject/Services/QuestService/QuestObjectives.cs:14]: Need to rename first second third data if it’s holding something precise

**Comment by Fizo55** [comment]
Would need too much code review right now, closed.

**Pull request closed without merging**